swupd/manifest.go

```diff
@@ -603,6 +606,139 @@ func (m *Manifest) hasUnsupportedTypeChanges() bool {
 	return false
 }
 
+// addAllManifestFiles creates a sorted list of file records for each manifest.
+func addAllManifestFiles(manifests []*Manifest, ui UpdateInfo, c config, numWorkers int) error {
+	var wg sync.WaitGroup
+	if numWorkers <= 0 {
+		numWorkers = runtime.NumCPU()
+	}
+	if numWorkers > len(ui.bundles) {
+		numWorkers = len(ui.bundles)
+	}
+	wg.Add(numWorkers)
+
+	mCh := make(chan *Manifest)
+	errorChan := make(chan error, numWorkers)
+	defer close(errorChan)
+
+	manifestWorker := func() {
+		defer wg.Done()
+		for m := range mCh {
+			if err := m.addManifestFiles(ui, c); err != nil {
+				errorChan <- err
+				return
+			}
+		}
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go manifestWorker()
+	}
+
+	var err error
+	for _, m := range manifests {
+		select {
+		case mCh <- m:
+		case err = <-errorChan:
+			close(mCh)
+			return err
+		}
+	}
+
+	close(mCh)
+	wg.Wait()
+
+	if err == nil && len(errorChan) > 0 {
+		err = <-errorChan
+	}
+
+	return err
+}
+
+// addManifestFiles creates a sorted list of file records for a manifest.
+// Files in the manifest's bundle info file that exist in included bundles
+// will not be added.
+func (m *Manifest) addManifestFiles(ui UpdateInfo, c config) error {
+	if m.Name == "full" {
+		chroot := filepath.Join(c.imageBase, fmt.Sprint(ui.version), "full")
+		if err := m.addFilesFromChroot(chroot, ""); err != nil {
+			return err
+		}
+	} else {
+		// Add files to manifest that do not exist in included bundles.
+		for f := range m.BundleInfo.Files {
+			if m.isFileIncluded(f) == false {
```

Comment:
The isFileIncluded re-calculates the recursive includes list for each file in the manifest. I am going to replace this function with a function that gets all recursive bundle includes. This can be calculated once and re-used.